src/send_kcidb.py

```diff
@@ -8,126 +8,116 @@
 
-    def __call__(self, configs, args):
-        db_config = configs['db_configs'][args.db_config]
-        api_token = os.getenv('API_TOKEN')
-        db = kernelci.db.get_db(db_config, api_token)
-
-        if not os.getenv('GOOGLE_APPLICATION_CREDENTIALS'):
-            self._logger.log_message(logging.ERROR,
-                                     "No GOOGLE_APPLICATION_CREDENTIALS \
-environment variable")
-            return False
-
-        topic_name = os.getenv('KCIDB_TOPIC_NAME')
-        if not topic_name:
-            self._logger.log_message(logging.ERROR, "No KCIDB_TOPIC_NAME \
-environment variable")
-            return False
-
-        project_id = os.getenv('KCIDB_PROJECT_ID')
-        if not project_id:
-            self._logger.log_message(logging.ERROR, "No KCIDB_PROJECT_ID \
-environment variable")
-            return False
-
-        client = Client(project_id=project_id, topic_name=topic_name)
-        if client is None:
-            self._logger.log_message(logging.ERROR, "Failed to create client \
-connection to KCIDB")
-            return False
-
-        sub_id = db.subscribe_node_channel(filters={
+    def _setup(self, args):
+        self._client = self._kcidb_connect()
+        return self._db.subscribe_node_channel(filters={
             'name': 'checkout',
             'state': 'done',
         })
-        self._logger.log_message(logging.INFO, "Listening for events... ")
-        self._logger.log_message(logging.INFO, "Press Ctrl-C to stop.")
-        sys.stdout.flush()
 
-        tz_utc = datetime.timezone(datetime.timedelta(hours=0))
-
-        try:
-            while True:
-                node = db.receive_node(sub_id)
-                self._logger.log_message(logging.INFO, f"Submitting node to \
-KCIDB: {node['_id']}")
-                sys.stdout.flush()
-
-                created_time = datetime.datetime.fromisoformat(node["created"])
-                if not created_time.tzinfo:
-                    created_time = datetime.datetime.fromtimestamp(
-                        created_time.timestamp(), tz=tz_utc)
-                revision = {
-                    "builds": [],
-                    "checkouts": [{
-                        "id": f"kernelci:{node['_id']}",
-                        "origin": "kernelci",
-                        "tree_name": node["revision"]["tree"],
-                        "git_repository_url": node["revision"]["url"],
-                        "git_commit_hash": node["revision"]["commit"],
-                        "git_repository_branch": node["revision"]["branch"],
-                        "start_time": created_time.isoformat(),
-                        "patchset_hash": "",
-                    }],
-                    "tests": [],
-                    "version": {
-                        "major": 4,
-                        "minor": 0
-                    }
-                }
-                self.send_revision(client, revision)
+    def _stop(self, sub_id):
+        if sub_id:
```

Comment:
I guess the KCIDB Client object doesn't have any means of disconnecting?

If it's held in the context object then it will go out of scope when the loop exits and any clean-up will be done automatically then.

Reply:
I checked `kcidb.Client` implementation. It doesn't have any external method to disconnect.